src/pybind/rbd/c_rbd.pxd

```diff
@@ -525,7 +526,7 @@ cdef extern from "rbd/librbd.h" nogil:
     int rbd_snap_unprotect(rbd_image_t image, const char *snap_name)
     int rbd_snap_is_protected(rbd_image_t image, const char *snap_name,
                               int *is_protected)
-    int rbd_snap_exists(rbd_image_t image, const char *snapname, bint *exists)
+    int rbd_snap_exists(rbd_image_t image, const char *snapname, libcpp.bool *exists)
```

Comment:
Makes sense, although it's clear that `libcpp` was intended only for C++ because it doesn't actually pull in `stdbool.h`.  Looks like someone else noticed this too: https://github.com/cython/cython/issues/5730.

I'm still confused about 

> Part of the GCC 14 preparations, but this looks like a real bug.

though.  Given that it's initialized (`bint _exists = False`, translated to `int <generated_name> = 0;`), that the API sets it to either `true` or `false`, and that we then do `bool(_exists != 0)`, what can go wrong?  Can you include the warning or error reported by GCC 14 in the commit message?

Reply:
I'm asking because if it's a real bug, we would need to backport this.

Reply:
The docs build involves cythonization for producing the API reference (see https://docs.ceph.com/en/latest/rbd/api/librbdpy/#module-rbd) and it's failing due to https://github.com/cython/cython/issues/5730:
```
  running bdist_wheel
  running build
  running build_ext
  building 'rbd' extension
  creating build
  creating build/temp.linux-x86_64-3.10
  x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/include -I/usr/include/python3.10 -c rbd.c -o build/temp.linux-x86_64-3.10/rbd.o
  rbd.c:4341:81: error: unknown type name 'bool'
   4341 | static int __pyx_f_3rbd_rbd_snap_exists(__pyx_t_3rbd_rbd_image_t, char const *, bool *); /*proto*/
        |                                                                                 ^~~~
  rbd.c:3805:1: note: 'bool' is defined in header '<stdbool.h>'; did you forget to '#include <stdbool.h>'?
   3804 | #include <stdalign.h>
    +++ |+#include <stdbool.h>
   3805 | #endif
```
I think the real build ends up using the C++ compiler so there we get away with it.

Would replacing `cimport libcpp` with
```
cdef extern from *:
    ctypedef bint _Bool
```
and replacing `libcpp.bool` with `_Bool` work?  It seems like that should achieve the same result.

The alternative is digging though the build scripts...

Reply:
> C++ does not have `_Bool`, so if you are indeed compiling with C++, it's not going to work.

Yup, I meant to edit that but got distracted.  It looks like it's compiled as C and the difference is caused by the
```
IF BUILD_DOC:
    include "mock_rbd.pxi"
ELSE:
    from c_rbd cimport *
    cimport rados
```
bit in `src/pybind/rbd/rbd.pyx`.  Adding
```
cdef extern from "<stdbool.h>":
    pass
```
just to `src/pybind/rbd/mock_rbd.pxi` should work around it.  Feel free to force push that and we will see.

Reply:
... although I would prefer my `_Bool` approach, just to avoid referencing `libcpp` and then having to work around it being intended for C++.

Reply:
> we could add:
> 
> ```cython
> cdef extern from "<stdbool.h>":
>     ctypedef bint bool
> ```

I think that would mess with some `bool(...)` syntax in Python code, like:

https://github.com/ceph/ceph/blob/86bb77eb9633bfd002e73b5e58b863bc2d0df594/src/pybind/rbd/rbd.pyx#L2598

If it's easy to adjust, let's do it.

Reply:
Also, as I look at this more, in particular `src/pybind/rbd/c_rbd.pxd`, there are more instances of `bint` being used in place of a `bool`, both in structs:
```
    ctypedef struct rbd_snap_mirror_namespace_t:
        rbd_snap_mirror_state_t state
        size_t mirror_peer_uuids_count
        char *mirror_peer_uuids
        bint complete
        char *primary_mirror_uuid
        uint64_t primary_snap_id
        uint64_t last_copied_object_number
```
for
```
typedef struct {
  rbd_snap_mirror_state_t state;
  size_t mirror_peer_uuids_count;
  char *mirror_peer_uuids;
  bool complete;
  char *primary_mirror_uuid;
  uint64_t primary_snap_id;
  uint64_t last_copied_object_number;
} rbd_snap_mirror_namespace_t;
```
and as a non-pointer parameter:
```
    int rbd_mirror_image_disable(rbd_image_t image, bint force)
```
for
```
CEPH_RBD_API int rbd_mirror_image_disable(rbd_image_t image, bool force);
```
We probably need to clean up all of `bint` usage here...

Reply:
Can you move the RBD one to `.pyx` or the RGW one to `.pxi` so that we are consistent?